network/src/protocols/network/mod.rs

```diff
@@ -172,22 +172,67 @@ pub trait NewNetworkEvents {
     fn new(
         peer_mgr_notifs_rx: aptos_channel::Receiver<(PeerId, ProtocolId), PeerManagerNotification>,
         connection_notifs_rx: aptos_channel::Receiver<PeerId, ConnectionNotification>,
+        max_parallel_deserialization_tasks: Option<usize>,
     ) -> Self;
 }
 
-impl<TMessage: Message> NewNetworkEvents for NetworkEvents<TMessage> {
+impl<TMessage: Message + Send + 'static> NewNetworkEvents for NetworkEvents<TMessage> {
     fn new(
         peer_mgr_notifs_rx: aptos_channel::Receiver<(PeerId, ProtocolId), PeerManagerNotification>,
         connection_notifs_rx: aptos_channel::Receiver<PeerId, ConnectionNotification>,
+        max_parallel_deserialization_tasks: Option<usize>,
     ) -> Self {
-        let data_event_stream = peer_mgr_notifs_rx.filter_map(
-            peer_mgr_notif_to_event
-                as fn(PeerManagerNotification) -> future::Ready<Option<Event<TMessage>>>,
+        // Create a channel for deserialized messages
+        let (deserialized_message_sender, deserialized_message_receiver) = aptos_channel::new(
+            QueueStyle::FIFO,
+            MAX_DESERIALIZATION_QUEUE_SIZE_PER_PEER,
+            None,
         );
+
+        // Deserialize the peer manager notifications in parallel (for each
+        // network application) and send them to the receiver. Note: this
+        // may cause out of order message delivery, but applications
+        // should already be handling this.
+        Handle::current().spawn(async move {
+            peer_mgr_notifs_rx
+                .for_each_concurrent(
+                    max_parallel_deserialization_tasks,
+                    move |peer_manager_notification| {
+                        // Get the peer ID for the notification
+                        let deserialized_message_sender = deserialized_message_sender.clone();
+                        let peer_id_for_notification = peer_manager_notification.get_peer_id();
+
+                        // Spawn a new blocking task to deserialize the message
+                        Handle::current()
```

Comment:
nit: there's a shortcut for this - tokio::task::spawn_blocking

Reply:
Done, thanks! 😄